File: src/feeds/coinbase-ws.ts
/**
 * Coinbase Advanced Trade WebSocket feed.
 *
 * Manages two subscriptions:
 *  1. ticker   — real-time price ticks for all watched symbols
 *  2. level2   — order book updates (bids/asks) for L2 imbalance strategy
 *
 * Reconnection strategy:
 *  - Exponential backoff: 1s, 2s, 4s, 8s, 16s, capped at 30s
 *  - Resubscribes automatically on reconnect
 *  - Logs each reconnection attempt with backoff details
 *
 * Message routing:
 *  - Ticker  → onTick(symbol, price, volume24h)
 *  - Level2  → onBookUpdate(symbol, bids, asks)
 */

import WebSocket from 'ws';
import { log } from '../storage/database.js';

// ─── Types ────────────────────────────────────────────────────────────────────

type TickCallback     = (symbol: string, price: number, volume24h: number) => void;
type BookCallback     = (symbol: string, bids: { price: number; size: number }[], asks: { price: number; size: number }[]) => void;
type ConnectionStatus = 'connecting' | 'connected' | 'disconnected';

interface CoinbaseTicker {
  type: 'ticker';
  product_id: string;
  price: string;
  volume_24h: string;
  best_bid: string;
  best_ask: string;
  time: string;
}

interface CoinbaseL2Update {
  type: 'l2update';
  product_id: string;
  changes: Array<['buy' | 'sell', string, string]>; // [side, price, size]
  time: string;
}

interface CoinbaseSubscriptionsMsg {
  type: 'subscriptions';
  channels: Array<{ name: string; product_ids: string[] }>;
}

type CoinbaseMessage = CoinbaseTicker | CoinbaseL2Update | CoinbaseSubscriptionsMsg | { type: string };

const WS_URL = 'wss://advanced-trade-ws.coinbase.com';
const MAX_BACKOFF_MS = 30_000;

// ─── In-memory order book ─────────────────────────────────────────────────────

const bookState = new Map<string, { bids: Map<string, number>; asks: Map<string, number> }>();

function getBook(productId: string) {
  if (!bookState.has(productId)) {
    bookState.set(productId, { bids: new Map(), asks: new Map() });
  }
  return bookState.get(productId)!;
}

function applyL2Update(update: CoinbaseL2Update, onBook: BookCallback) {
  const book = getBook(update.product_id);
  for (const [side, priceStr, sizeStr] of update.changes) {
    const price = parseFloat(priceStr);
    const size = parseFloat(sizeStr);
    const map = side === 'buy' ? book.bids : book.asks;
    if (size === 0) {
      map.delete(priceStr);
    } else {
      map.set(priceStr, size);
    }
  }

  // Emit sorted top-20 each side
  const symbol = update.product_id.replace('-USD', '');
  const bids = Array.from(book.bids.entries())
    .map(([p, s]) => ({ price: parseFloat(p), size: s }))
    .sort((a, b) => b.price - a.price)
    .slice(0, 20);
  const asks = Array.from(book.asks.entries())
    .map(([p, s]) => ({ price: parseFloat(p), size: s }))
    .sort((a, b) => a.price - b.price)
    .slice(0, 20);

  onBook(symbol, bids, asks);
}

// ─── Connection manager ───────────────────────────────────────────────────────

export class CoinbaseWebSocket {
  private ws: WebSocket | null = null;
  private productIds: string[] = [];
  private onTick: TickCallback;
  private onBook: BookCallback;
  private status: ConnectionStatus = 'disconnected';
  private reconnectAttempts = 0;
  private reconnectTimer: ReturnType<typeof setTimeout> | null = null;

  constructor(productIds: string[], onTick: TickCallback, onBook: BookCallback) {
    this.productIds = productIds;
    this.onTick = onTick;
    this.onBook = onBook;
  }

  connect(): void {
    if (this.status === 'connected' || this.status === 'connecting') return;
    this.status = 'connecting';
    log('info', `Coinbase WS connecting (attempt ${this.reconnectAttempts + 1})...`);

    const ws = new WebSocket(WS_URL);
    this.ws = ws;

    ws.on('open', () => {
      this.status = 'connected';
      this.reconnectAttempts = 0;
      log('info', `Coinbase WS connected — subscribing to ${this.productIds.length} products`);
      this.subscribe();
    });

    ws.on('message', (raw: Buffer) => {
      try {
        const msg = JSON.parse(raw.toString()) as CoinbaseMessage;
        this.handleMessage(msg);
      } catch { /* malformed JSON — ignore */ }
    });

    ws.on('error', (err) => {
      log('warn', `Coinbase WS error: ${err.message}`);
    });

    ws.on('close', () => {
      this.status = 'disconnected';
      log('warn', 'Coinbase WS disconnected — scheduling reconnect');
      this.scheduleReconnect();
    });
  }

  private subscribe(): void {
    if (!this.ws || this.ws.readyState !== WebSocket.OPEN) return;
    this.ws.send(JSON.stringify({
      type: 'subscribe',
      product_ids: this.productIds,
      channels: ['ticker', 'level2'],
    }));
  }

  private handleMessage(msg: CoinbaseMessage): void {
    switch (msg.type) {
      case 'ticker': {
        const tick = msg as CoinbaseTicker;
        const symbol = tick.product_id.replace('-USD', '');
        const price = parseFloat(tick.price);
        const volume = parseFloat(tick.volume_24h);
        if (!isNaN(price) && price > 0) {
          this.onTick(symbol, price, volume);
        }
        break;
      }
      case 'l2update': {
        applyL2Update(msg as CoinbaseL2Update, this.onBook);
        break;
      }
      case 'subscriptions': {
        log('info', 'Coinbase WS subscriptions confirmed');
        break;
      }
      // Ignore heartbeat, snapshot, etc.
    }
  }

  private scheduleReconnect(): void {
    if (this.reconnectTimer) return;
    const backoffMs = Math.min(MAX_BACKOFF_MS, 1_000 * Math.pow(2, this.reconnectAttempts));
    this.reconnectAttempts++;
    log('info', `Coinbase WS reconnecting in ${backoffMs}ms (attempt ${this.reconnectAttempts})`);
    this.reconnectTimer = setTimeout(() => {
      this.reconnectTimer = null;
      this.connect();
    }, backoffMs);
  }

  disconnect(): void {
    if (this.reconnectTimer) {
      clearTimeout(this.reconnectTimer);
      this.reconnectTimer = null;
    }
    this.ws?.close();
    this.status = 'disconnected';
  }

  isConnected(): boolean {
    return this.status === 'connected';
  }

  updateProducts(productIds: string[]): void {
    this.productIds = productIds;
    if (this.status === 'connected') this.subscribe();
  }
}

File: src/feeds/coinbase_ws.py
"""Coinbase Advanced Trade WebSocket feed using the official SDK."""

import json
import threading
import time
from typing import Callable, Optional

from src.config import env
from src.storage.database import log

BOOK_STALE_S = 60

_book_state: dict[str, dict] = {}
_book_lock = threading.Lock()


def _get_book(product_id: str) -> dict:
    if product_id not in _book_state:
        _book_state[product_id] = {"bids": {}, "asks": {}, "updated_at": time.time()}
    return _book_state[product_id]


def purge_stale_books(max_age_s: float = BOOK_STALE_S) -> int:
    """Remove book entries older than max_age_s. Returns number of purged products."""
    now = time.time()
    purged = 0
    with _book_lock:
        stale_keys = [
            k for k, v in _book_state.items()
            if now - v.get("updated_at", 0) > max_age_s
        ]
        for k in stale_keys:
            del _book_state[k]
            purged += 1
    if purged:
        log("info", f"Purged {purged} stale book entries")
    return purged


class CoinbaseWebSocket:
    def __init__(
        self,
        product_ids: list[str],
        on_tick: Callable[[str, float, float], None],
        on_book: Callable[[str, list, list], None],
    ):
        self.product_ids = product_ids
        self.on_tick = on_tick
        self.on_book = on_book
        self._client = None
        self._status = "disconnected"
        self._thread: Optional[threading.Thread] = None
        self._tick_count = 0
        self._last_tick_time = time.time()
        self._stop_event = threading.Event()
        self._last_vol_24h: dict[str, float] = {}  # track 24h volume to compute tick deltas

    def connect(self) -> None:
        if self._status in ("connected", "connecting"):
            return
        self._status = "connecting"
        self._stop_event.clear()

        try:
            from coinbase.websocket import WSClient
        except ImportError:
            log("error", "coinbase-advanced-py not installed — pip install coinbase-advanced-py")
            return

        api_key = env.coinbase_api_key or ""
        api_secret = env.coinbase_api_secret or ""

        def _make_ws_client():
            """Create a WSClient instance (shared between initial connect and reconnect)."""
            if api_key and api_secret:
                return WSClient(
                    api_key=api_key,
                    api_secret=api_secret,
                    on_message=self._on_message,
                    on_close=self._on_close,
                    retry=False,
                )
            return WSClient(
                on_message=self._on_message,
                on_close=self._on_close,
                retry=False,
            )

        self._client = _make_ws_client()

        def _run():
            from coinbase.websocket import WSClientConnectionClosedException, WSClientException

            while not self._stop_event.is_set():
                try:
                    self._client.open()
                    self._status = "connected"
                    log("info", f"Coinbase WS connected — subscribing to {len(self.product_ids)} products")

                    self._client.subscribe(
                        product_ids=self.product_ids,
                        channels=["ticker", "heartbeats"],
                    )

                    log("info", "Coinbase WS subscriptions sent")

                    while not self._stop_event.is_set() and self._status == "connected":
                        self._client.sleep_with_exception_check(sleep=5)

                except WSClientConnectionClosedException:
                    log("warn", "Coinbase WS connection closed — reconnecting in 5s")
                except WSClientException as e:
                    log("warn", f"Coinbase WS error: {e} — reconnecting in 5s")
                except Exception as e:
                    log("error", f"Coinbase WS connection failed: {e} — reconnecting in 5s")

                self._status = "disconnected"
                if self._stop_event.is_set():
                    break

                if self._stop_event.wait(timeout=5):
                    break

                # Create a fresh client for reconnection
                try:
                    self._client = _make_ws_client()
                except Exception as e:
                    log("error", f"Failed to create WS client: {e}")
                    if self._stop_event.wait(timeout=10):
                        break

        self._thread = threading.Thread(target=_run, daemon=True, name="coinbase-ws")
        self._thread.start()

    def _on_message(self, raw: str) -> None:
        try:
            msg = json.loads(raw)
        except (json.JSONDecodeError, TypeError):
            return

        channel = msg.get("channel")
        events = msg.get("events", [])

        for event in events:
            if channel == "ticker":
                tickers = event.get("tickers", [])
                for t in tickers:
                    product_id = t.get("product_id", "")
                    symbol = product_id.replace("-USD", "")
                    try:
                        price = float(t["price"])
                        # Use per-tick volume if available, fallback to 24h volume
                        # Coinbase ticker provides volume_24_h (cumulative) — not per-tick
                        # Use the difference from last known 24h volume as proxy for tick volume
                        raw_vol_24h = float(t.get("volume_24_h", 0))
                        last_vol = self._last_vol_24h.get(symbol, raw_vol_24h)
                        volume = max(0, raw_vol_24h - last_vol) if last_vol > 0 else 0
                        self._last_vol_24h[symbol] = raw_vol_24h
                        if price > 0:
                            self.on_tick(symbol, price, volume)
                            self._tick_count += 1
                            self._last_tick_time = time.time()
                            if self._tick_count == 1:
                                log("info", f"First tick received: {symbol} @ ${price:,.2f}")
                    except Exception as e:
                        if self._tick_count < 5:
                            log("error", f"Tick handler error for {symbol}: {type(e).__name__}: {e}")
                        pass

            elif channel == "l2_data":
                product_id = event.get("product_id", "")
                updates = event.get("updates", [])
                if not updates:
                    continue
                with _book_lock:
                    book = _get_book(product_id)
                    for update in updates:
                        try:
                            side = update.get("side", "")
                            price_str = update.get("price_level", "")
                            size = float(update.get("new_quantity", 0))
                            book_side = book["bids"] if side == "bid" else book["asks"]
                            if size == 0:
                                book_side.pop(price_str, None)
                            else:
                                book_side[price_str] = size
                        except (ValueError, TypeError):
                            pass
                    book["updated_at"] = time.time()
                    bids_snapshot = list(book["bids"].items())
                    asks_snapshot = list(book["asks"].items())

                symbol = product_id.replace("-USD", "")
                bids = sorted(
                    [{"price": float(p), "size": s} for p, s in bids_snapshot],
                    key=lambda x: x["price"], reverse=True,
                )[:20]
                asks = sorted(
                    [{"price": float(p), "size": s} for p, s in asks_snapshot],
                    key=lambda x: x["price"],
                )[:20]
                self.on_book(symbol, bids, asks)

    def _on_close(self) -> None:
        was_connected = self._status == "connected"
        self._status = "disconnected"
        if was_connected:
            log("warn", "WebSocket disconnected — reconnecting")

    def check_health(self, max_silence_s: float = 30.0) -> bool:
        """Check if WS is receiving ticks. Force-reconnect if stale."""
        if self._status != "connected":
            return False
        silence = time.time() - self._last_tick_time
        if silence > max_silence_s:
            log("warn", f"WS tick silence for {silence:.0f}s — force reconnecting")
            self.force_reconnect()
            return False
        return True

    def force_reconnect(self) -> None:
        """Kill the current connection and reconnect."""
        self._status = "disconnected"
        if self._client:
            try:
                self._client.close()
            except Exception:
                pass
        # The _run loop will detect the disconnection and reconnect automatically
        self._tick_count = 0
        self._last_tick_time = time.time()

    def disconnect(self) -> None:
        self._stop_event.set()
        if self._client:
            try:
                self._client.close()
            except Exception:
                pass
        self._status = "disconnected"

    def is_connected(self) -> bool:
        return self._status == "connected"

    def update_products(self, product_ids: list[str]) -> None:
        old = set(self.product_ids)
        new = set(product_ids)
        self.product_ids = product_ids

        if not self._client or self._status != "connected":
            return

        unsub = list(old - new)
        sub = list(new - old)
        if unsub:
            try:
                self._client.unsubscribe(product_ids=unsub, channels=["ticker"])
            except Exception:
                pass
        if sub:
            try:
                self._client.subscribe(product_ids=sub, channels=["ticker"])
            except Exception:
                pass
